File: app/src/main/java/univ/orleans/ttl/isokachallenge/orm/Tables.java
package univ.orleans.ttl.isokachallenge.orm;

public class Tables {

    //  ###########################
    //  #   DATABASE INFORMATION  #
    //  ###########################

    public static final String DB_NAME = "isoka_db";
    public static final int DB_VERSION = 1;

    public static final String DB_LOG = "isoka_db_log";

    //  #################
    //  #   USER TABLE  #
    //  #################

    public static final String USER_TABLE = "USER";
    public static final String USER_NAME = "u_username";
    public static final int USER_NAME_INDEX = 0;
    public static final String USER_DATE = "u_date";
    public static final int USER_DATE_INDEX = 1;

    public static final String USER_CREATE = "CREATE TABLE " + USER_TABLE + "(" + USER_NAME + " VARCHAR(255) PRIMARY KEY NOT NULL," + USER_DATE + " DATE NOT NULL);";

    //  ####################
    //  #   DRAWING TABLE  #
    //  ####################

    public static final String DRAWING_TABLE = "DRAWING";
    public static final String DRAWING_ID = "d_id";
    public static final int DRAWING_ID_INDEX = 0;
    public static final String DRAWING_LINK = "d_link";
    public static final int DRAWING_LINK_INDEX = 1;
    public static final String DRAWING_DATE = "d_date";
    public static final int DRAWING_DATE_INDEX = 2;

    public static final String DRAWING_CREATE = "CREATE TABLE " + DRAWING_TABLE + "("
            + DRAWING_ID + " INTEGER PRIMARY KEY NOT NULL,"
            + DRAWING_LINK + " VARCHAR(512) NOT NULL,"
            + DRAWING_DATE + " DATE NOT NULL);";

    //  ######################
    //  #   CHALLENGE TABLE  #
    //  ######################

    public static final String CHALLENGE_TABLE = "CHALLENGE";
    public static final String CHALLENGE_ID = "c_id";
    public static final int CHALLENGE_ID_INDEX = 0;
    public static final String CHALLENGE_NAME = "c_name";
    public static final int CHALLENGE_NAME_INDEX = 1;
    public static final String CHALLENGE_TYPE = "c_type";
    public static final int CHALLENGE_TYPE_INDEX = 2;
    public static final String CHALLENGE_THEME = "c_theme";
    public static final int CHALLENGE_THEME_INDEX = 3;
    public static final String CHALLENGE_DURATION = "c_duration";
    public static final int CHALLENGE_DURATION_INDEX = 4;
    public static final String CHALLENGE_TIMER = "c_timer";
    public static final int CHALLENGE_TIMER_INDEX = 5;
    public static final String CHALLENGE_DESCRIPTION = "c_desc";
    public static final int CHALLENGE_DESCRIPTION_INDEX = 6;

    public static final String CHALLENGE_CREATE = "CREATE TABLE " + CHALLENGE_TABLE + "("
            + CHALLENGE_ID + " INTEGER PRIMARY KEY NOT NULL,"
            + CHALLENGE_NAME + " VARCHAR(255) NOT NULL,"
            + CHALLENGE_TYPE + " BOOLEAN NOT NULL,"
            + CHALLENGE_THEME + " VARCHAR(255) NOT NULL,"
            + CHALLENGE_DURATION + " DATE NOT NULL,"
            + CHALLENGE_TIMER + " INTEGER NOT NULL,"
            + CHALLENGE_DESCRIPTION + " VARCHAR(255) NOT NULL);";

    //  ##########################
    //  #   PARTICIPATION TABLE  #
    //  ##########################

    public static final String PARTICIPATION_TABLE = "PARTICIPATION";
    public static final String PARTICIPATION_USER_ID = "p_u_id";
    public static final int PARTICIPATION_USER_ID_INDEX = 0;
    public static final String PARTICIPATION_DRAWING_ID = "p_d_id";
    public static final int PARTICIPATION_DRAWING_ID_INDEX = 1;
    public static final String PARTICIPATION_CHALLENGE_ID = "p_c_id";
    public static final int PARTICIPATION_CHALLENGE_ID_INDEX = 2;
    public static final String PARTICIPATION_IS_CREATOR = "p_is_creator";
    public static final int PARTICIPATION_IS_CREATOR_INDEX = 3;
    public static final String PARTICIPATION_VOTES = "p_votes";
    public static final int PARTICIPATION_VOTES_INDEX = 4;

    public static final String PARTICIPATION_CREATE = "CREATE TABLE " + PARTICIPATION_TABLE + "("
            + PARTICIPATION_USER_ID + " VARCHAR(255) NOT NULL,"
            + PARTICIPATION_DRAWING_ID + " INTEGER NOT NULL,"
            + PARTICIPATION_CHALLENGE_ID + " INTEGER NOT NULL,"
            + PARTICIPATION_IS_CREATOR + " BOOLEAN NOT NULL,"
            + PARTICIPATION_VOTES + " INTEGER DEFAULT 0,"
            + "FOREIGN KEY(" + PARTICIPATION_USER_ID + ") REFERENCES " + USER_TABLE + "(" + USER_NAME + ") ON DELETE CASCADE,"
            + "FOREIGN KEY(" + PARTICIPATION_DRAWING_ID + ") REFERENCES " + DRAWING_TABLE + "( " + DRAWING_ID + ") ON DELETE CASCADE,"
            + "FOREIGN KEY(" + PARTICIPATION_CHALLENGE_ID + ") REFERENCES " + CHALLENGE_TABLE + "(" + CHALLENGE_ID + ") ON DELETE CASCADE,"
            + "PRIMARY KEY(" + PARTICIPATION_USER_ID + "," + PARTICIPATION_DRAWING_ID + "," + PARTICIPATION_CHALLENGE_ID + "));";

    //  ############
    //  #   OTHER  #
    //  ############

    public static final String DROP = "DROP TABLE IF EXISTS " + PARTICIPATION_TABLE + ";"
            + "DROP TABLE IF EXISTS " + USER_TABLE + ";"
            + "DROP TABLE IF EXISTS " + DRAWING_TABLE + ";"
            + "DROP TABLE IF EXISTS " + CHALLENGE_TABLE + ";";

    //  ############
    //  #   OTHER  #
    //  ############

    public static final String OPERATOR_EQ = " = ";
    public static final String OPERATOR_DIFF = " <> ";
    public static final String OPERATOR_INF = " < ";
    public static final String OPERATOR_INF_EQ = " <= ";
    public static final String OPERATOR_SUP = " > ";
    public static final String OPERATOR_SUP_EQ = " >= ";
    public static final String OPERATOR_LIKE = " LIKE ";
}

File: app/src/main/java/univ/orleans/ttl/isokachallenge/orm/Tables.kt
package univ.orleans.ttl.isokachallenge.orm

//  ###########################
//  #   DATABASE INFORMATION  #
//  ###########################

const val DB_NAME: String = "isoka_db"
const val DB_VERSION: Int = 1

const val DB_LOG: String = "isoka_db_log"

//  #################
//  #   USER TABLE  #
//  #################

const val USER_TABLE: String = "USER"
const val USER_NAME: String = "u_username"
const val USER_DATE: String = "u_date"

const val USER_CREATE: String = "CREATE TABLE $USER_TABLE($USER_NAME VARCHAR(255) PRIMARY KEY, $USER_DATE DATE NOT NULL);"

//  ####################
//  #   DRAWING TABLE  #
//  ####################

const val DRAWING_TABLE: String = "DRAWING"
const val DRAWING_ID: String = "d_id"
const val DRAWING_LINK: String = "d_link"
const val DRAWING_DATE: String = "d_date"

const val DRAWING_CREATE: String = "CREATE TABLE $DRAWING_TABLE($DRAWING_ID INTEGER PRIMARY KEY, $DRAWING_LINK VARCHAR(512) NOT NULL, $DRAWING_DATE DATE NOT NULL);"

//  ######################
//  #   CHALLENGE TABLE  #
//  ######################

const val CHALLENGE_TABLE: String = "CHALLENGE"
const val CHALLENGE_ID: String = "c_id"
const val CHALLENGE_NAME: String = "c_name"
const val CHALLENGE_TYPE: String = "c_type"
const val CHALLENGE_THEME: String = "c_theme"
const val CHALLENGE_DURATION: String = "c_duration"
const val CHALLENGE_TIMER: String = "c_timer"

const val CHALLENGE_CREATE: String = "CREATE TABLE $CHALLENGE_TABLE($CHALLENGE_ID INTEGER PRIMARY KEY, $CHALLENGE_NAME VARCHAR(255) NOT NULL, $CHALLENGE_TYPE BOOLEAN NOT NULL, $CHALLENGE_THEME VARCHAR(255) NOT NULL, $CHALLENGE_DURATION INTEGER NOT NULL, $CHALLENGE_TIMER INTEGER NOT NULL);"

//  ##########################
//  #   PARTICIPATION TABLE  #
//  ##########################

const val PARTICIPATION_TABLE: String = "PARTICIPATION"
const val PARTICIPATION_CHALLENGE_ID: String = "p_c_id"
const val PARTICIPATION_USER_ID: String = "p_u_id"
const val PARTICIPATION_DRAWING_ID: String = "p_d_id"
const val PARTICIPATION_IS_CREATOR: String = "p_is_creator"
const val PARTICIPATION_VOTES: String = "p_votes"

const val PARTICIPATION_CREATE: String = "CREATE TABLE $PARTICIPATION_TABLE($PARTICIPATION_USER_ID VARCHAR(255), $PARTICIPATION_DRAWING_ID INTEGER, $PARTICIPATION_CHALLENGE_ID INTEGER, $PARTICIPATION_IS_CREATOR BOOLEAN NOT NULL, $PARTICIPATION_VOTES INTEGER DEFAULT 0, FOREIGN KEY($PARTICIPATION_USER_ID) REFERENCES $USER_TABLE($USER_NAME) ON DELETE CASCADE, FOREIGN KEY($PARTICIPATION_DRAWING_ID) REFERENCES $DRAWING_TABLE($DRAWING_ID) ON DELETE CASCADE, FOREIGN KEY($PARTICIPATION_CHALLENGE_ID) REFERENCES $CHALLENGE_TABLE($CHALLENGE_ID) ON DELETE CASCADE, PRIMARY KEY($PARTICIPATION_USER_ID, $PARTICIPATION_DRAWING_ID, $PARTICIPATION_CHALLENGE_ID));"

//  ############
//  #   OTHER  #
//  ############

const val DROP: String = "DROP TABLE IF EXISTS $PARTICIPATION_TABLE; DROP TABLE IF EXISTS $USER_TABLE; DROP TABLE IF EXISTS $DRAWING_TABLE; DROP TABLE IF EXISTS $CHALLENGE_TABLE;"